### PR TITLE
[MRG+1] Fix show_fiff.

### PR DESCRIPTION
**Pull request by jaeilepp**
Closes https://github.com/mne-tools/mne-python/issues/4095

Are we happy with this?

### CODE REVIEW

**Comment by agramfort** [comment]
travis is not happy

File: mne/io/open.py [hunk 1]
```diff
@@ -236,6 +236,8 @@ def _show_tree(fid, tree, indent, level, read_limit, max_str):
                 postpend = ''
                 # print tag data nicely
                 if tag.data is not None:
+                    if isinstance(tag.data, unicode):
+                        tag.data = tag.data.encode('utf-8')
                     postpend = ' = ' + str(tag.data)[:max_str]
```
Comment (Eric89GXL) [inline mne/io/open.py]: There should be a better / more uniform way to do this (Py2 and Py3 compatible) rather than doing a conditional

**Comment by agramfort** [comment]
well it don't get a crash on master on this file and neither with this PR.

now looking at the output of 

mne show_fiff ~/Downloads/showfiff_breaks.fif

vs C version

mne_show_fiff --in ~/Downloads/showfiff_breaks.fif

the Python output is much less readable.

@jaeilepp can you have a look?

**Comment by Eric89GXL** [comment]
> well it don't get a crash on master on this file and neither with this PR.

Probably because you are on Python 3, where `str` is already `unicode`. I get the crash on Python 2.

File: mne/commands/mne_show_fiff.py [hunk 2]
```diff
@@ -21,11 +21,6 @@ def run():
     if len(args) != 1:
         parser.print_help()
         sys.exit(1)
-    # This works around an annoying bug on Windows for show_fiff, see:
-    # https://pythonhosted.org/kitchen/unicode-frustrations.html
-    if int(sys.version[0]) < 3:
-        UTF8Writer = codecs.getwriter('utf8')
-        sys.stdout = UTF8Writer(sys.stdout)
```
Comment (jaeilepp) [inline mne/commands/mne_show_fiff.py]: I don't know if this fixes the readability, but this is the only way I'm able to use this command with the problematic file. All the other configurations seem to fail.

Comment (agramfort) [inline mne/commands/mne_show_fiff.py]: this was certainly added for a reason :( I am not confident removing it like this...

**Comment by agramfort** [comment]
+1 for MRG

@Eric89GXL ok for you?

File: mne/io/open.py [hunk 3]
```diff
@@ -255,7 +257,6 @@ def _show_tree(fid, tree, indent, level, read_limit, max_str):
                 out += [next_idt + prepend + str(k) + ' = ' +
                         '/'.join(this_type) + ' (' + str(size) + ')' +
                         postpend]
-                out[-1] = out[-1].replace('\n', u'¶')
```
Comment (Eric89GXL) [inline mne/io/open.py]: why this change? This line is necessary to prevent `\n` from breaking the output.

Comment (jaeilepp) [inline mne/io/open.py]: It won't work with the utf-8 encoding. The output looks good for me with this change.

Comment (Teekuningas) [inline mne/io/open.py]: You can use just '¶' (or maybe u'\xc2\xb6'.encode('utf-8') to be clear as the sky) instead of u'¶'.
If you use u'¶' it will try to decode out[-1] to unicode with ascii codec to make replace possible thus giving DecodeError.

Comment (jaeilepp) [inline mne/io/open.py]: But it didn't work on every platform. I didn't try every option you stated, though. But I don't see the problem with the current changes.

Comment (Teekuningas) [inline mne/io/open.py]: Given some assumptions they should be equivalent.. I wonder how it did work before as the string got encoded to utf-8 before also because the pilcrow is not in basic ascii.

Comment (jaeilepp) [inline mne/io/open.py]: I believe it was unicode before. I think optimally this would have to check all the possible platforms to work 100% of the time. Not worth the effort IMO.

Comment (Teekuningas) [inline mne/io/open.py]: Yes I think so too. I wonder if it would be safest to stay as unicode now too so that if someone is expecting it be unicode (like if someone has show_fiff(..).encode('utf-8')) it won't break?

Comment (Teekuningas) [inline mne/io/open.py]: Though that itself contradicts with the signature where you can specify your output to be "str".

**Comment by Eric89GXL** [comment]
This version has the problem that a data tag with a newline in it (usually
a string entry) will cause a newline while printing, which will break the
flow. That's why this code was there in the first place.

**Comment by jaeilepp** [comment]
So how about replacing it with an empty string ``out[-1] = out[-1].replace('\n', '')``?

**Comment by jaeilepp** [comment]
And how about having a param for ``show_fiff`` that allows fetching a particular value? Usually users are only interested in a particular field.

**Comment by agramfort** [comment]
sure

**Comment by Eric89GXL** [comment]
> So how about replacing it with an empty string out[-1] = out[-1].replace('\n', '')?

It's less satisfying than the nice ¶ symbol, but if it's causing unsolvable problems then sure.

What I don't get is that the strings themselves can have unicode (or even that symbol) in them, so our code won't work for those, right? So it seems potentially indicative of a potentially larger problem. As @Teekuningas says we should do everything here with `unicode`. Perhaps casting with `six.text_type`, which is `unicode` on Python2 and `str` on Python3 (where it is unicode anyway), would work.

**Comment by jaeilepp** [comment]
Now also has the fix for https://github.com/mne-tools/mne-python/issues/4129.
Basically the gof is broadcasted to make the ``Dipole`` compatible.

File: mne/commands/mne_show_fiff.py [hunk 4]
```diff
@@ -17,16 +17,20 @@ def run():
     """Run command."""
     parser = mne.commands.utils.get_optparser(
         __file__, usage='mne show_fiff <file>')
+    parser.add_option("-t", "--tag", dest="tag",
+                      help="provide information about this tag", metavar="TAG")
```
Comment (jaeilepp) [inline mne/commands/mne_show_fiff.py]: I updated the example.

File: mne/io/open.py [hunk 5]
```diff
@@ -203,21 +206,33 @@ def _find_type(value, fmts=['FIFF_'], exclude=['FIFF_UNIT']):
...
     # print block-level information
     out = [this_idt + str(int(tree['block'])) + ' = ' +
            '/'.join(_find_type(tree['block'], fmts=['FIFFB_']))]
+    tag_found = False
+    if tag_id is not None:
+        tag_id = int(tag_id)  # mne show_fiff passes a string
+        if out[0].strip().startswith(str(tag_id)):
+            tag_found = True
+    else:
+        tag_found = True
```
Comment (Eric89GXL) [inline mne/io/open.py]: This can be simplified as:
```
tag_found = False
if tag_id is None or out[0].strip().startswith(str(tag_id)):
    tag_found = True
```

File: mne/io/open.py [hunk 6]
```diff
@@ -258,8 +273,10 @@ def _show_tree(fid, tree, indent, level, read_limit, max_str):
                 out[-1] = out[-1].replace('\n', u'¶')
                 counter = 0
                 good = True
-
+    if tag_id is not None and not tag_found:
```
Comment (Eric89GXL) [inline mne/io/open.py]: this can also just be `if not tag_found` because the only time that happens is when `tag_id is not None`

File: mne/dipole.py [hunk 7]
```diff
@@ -1235,7 +1235,10 @@ def _concatenate_dipoles(dipoles):
         pos.append(dipole.pos)
         amplitude.append(dipole.amplitude)
         ori.append(dipole.ori)
-        gof.append(dipole.gof)
+        if dipole.gof.ndim == 0:
+            gof.append(np.repeat(dipole.gof, len(dipole.times)))
+        else:
+            gof.append(dipole.gof)
```
Comment (agramfort) [inline mne/dipole.py]: it's not the right fix. See #4135

**Comment by Eric89GXL** [comment]
LGTM +1 for merge

**Comment by agramfort** [comment]
we're not there yet:

```
$ mne_show_fiff --in ~/Downloads/showfiff_breaks.fif --tag 269

 269 = SSS components
$ mne show_fiff ~/Downloads/showfiff_breaks.fif --tag=269
502 = FIFFB_SSS_INFO
    269 = FIFF_SSS_COMPONENTS (380) = [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ... array size=95
    278 = FIFF_SSS_NFREE (4) = [70]
    243 = FIFF_HPI_FIT_GOOD_LIMIT (4) = [ 0.98000002]
    244 = FIFF_HPI_FIT_DIST_LIMIT (4) = [ 0.005]
```

**Comment by Eric89GXL** [comment]
I agree, it should look like (show nesting, but not neighbors):
```
502 = FIFFB_SSS_INFO
    269 = FIFF_SSS_COMPONENTS (380) = [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ... array size=95
```

**Comment by Eric89GXL** [comment]
Once this works can you give us the output for:
```
 mne show_fiff ~/Downloads/showfiff_breaks.fif --tag=269
```

**Comment by jaeilepp** [comment]
```
502 = FIFFB_SSS_INFO
    269 = FIFF_SSS_COMPONENTS (380) = [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 ... array size=95
```

**Comment by Eric89GXL** [comment]
LGTM, +1 for merge then

**Comment by Eric89GXL** [comment]
... and since @agramfort gave +1 already, in it goes. Thanks @jaeilepp
